llvm/labs/lab2/prokofev_kirill/ReplaceMulWithShiftPlug.cpp

```diff
@@ -0,0 +1,106 @@
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Pass.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+
+namespace {
+struct ReplaceMulWithShift : llvm::PassInfoMixin<ReplaceMulWithShift> {
+public:
+  llvm::PreservedAnalyses run(llvm::Function &Func,
+                              llvm::FunctionAnalysisManager &FAM) {
+    std::vector<llvm::Instruction *> toRemove;
+    bool changed = false;
+    for (llvm::BasicBlock &BB : Func) {
+      for (llvm::Instruction &Inst : BB) {
+        if (!llvm::BinaryOperator::classof(&Inst)) {
+          continue;
+        }
+        llvm::BinaryOperator *op = llvm::cast<llvm::BinaryOperator>(&Inst);
+        if (op->getOpcode() != llvm::Instruction::BinaryOps::Mul) {
+          continue;
+        }
+
+        llvm::Value *leftOper = op->getOperand(0);
+        llvm::Value *rightOper = op->getOperand(1);
+
+        int logVal1 = getLogBase2(leftOper);
+        int logVal2 = getLogBase2(rightOper);
+        if (logVal1 < logVal2) {
+          std::swap(logVal1, logVal2);
+          std::swap(leftOper, rightOper);
+        }
+
+        if (logVal1 > -1) {
+          llvm::Value *lg_val = llvm::ConstantInt::get(
+              llvm::IntegerType::get(Func.getContext(), 32),
+              llvm::APInt(32, logVal1));
+
+          llvm::Value *val = llvm::BinaryOperator::Create(
+              llvm::Instruction::Shl, rightOper, lg_val, op->getName(), op);
+
+          op->replaceAllUsesWith(val);
+          toRemove.push_back(op);
+          changed = true;
+        }
+      }
+      for (auto *I : toRemove) {
+        I->eraseFromParent();
+      }
+    }
+
+    if (changed)
+      return llvm::PreservedAnalyses::none();
+    return llvm::PreservedAnalyses::all();
+  }
+
+private:
+  int getLogBase2(llvm::Value *val) {
+    if (llvm::ConstantInt *CI = llvm::dyn_cast<llvm::ConstantInt>(val)) {
+      return CI->getValue().exactLogBase2();
+    }
+    if (auto *LI = llvm::dyn_cast<llvm::LoadInst>(val)) {
+      llvm::Value *Op = LI->getPointerOperand();
+      Op->reverseUseList();
+      llvm::StoreInst *StInst = nullptr;
+      for (auto *Inst : Op->users()) {
+        if (auto *SI = llvm::dyn_cast<llvm::StoreInst>(Inst)) {
+          StInst = SI;
+        }
+        if (Inst == LI) {
+          break;
```

Comment:
Why break? Why not to use `continue`?

Reply:
Because after finding a load instruction in a loop, it makes no sense to continue looking for a store instruction.